docs/ai/conceptual/understanding-tokens.md

```diff
@@ -0,0 +1,121 @@
+
+This text could be tokenized as:
+
+- I
+- heard
+- a
+- dog
+- bark
+- loudly
+- at
+- a
+- cat
+
+Only the first occurrence of the "a" token will be added to the model's vocabulary. All other occurrences can be represented by that first token.
+
+With a sufficiently large set of training text, a vocabulary of many thousands of tokens can be compiled.
+
+## How are tokens determined?
+
+The specific tokenization method varies by LLM. Some common tokenization methods include:
+
+- Word tokenization (text is split into individual words based on a delimiter)
+- Character tokenization (text is split into individual characters)
+- Subword tokenization (text is split into partial words or character sets)
+
+For example, the GPT models, developed by OpenAI, use _Byte-Pair Encoding_ (BPE) which is a type of subword tokenization. OpenAI provides [a tool to visualize how text will be tokenized](https://platform.openai.com/tokenizer).
+
+There are benefits and disadvantages to each tokenization method:
+
+**Smaller tokens (character or subword tokenization)**
+
+- _Pros_:
+  - Allows the model to handle a wider range of inputs, such as unknown words, typos, or complex syntax
+  - May allow the vocabulary size to be reduced, requiring less memory resources
+- _Cons_:
+  - A given text will be broken into more tokens requiring additional computational resources while processing
+  - Given a fixed token limit the maximum length of the model's input and output will be smaller
+
+**Larger tokens (word tokenization)**
+
+- _Pros_:
+  - A given text will be broken into fewer token requiring less computational resources while processing
+  - Given the same token limit the maximum size of the model's input and output will be larger
+- _Cons_:
+  - May cause an increased vocabulary size, requiring more memory resources
+  - Can limit the models ability to handle unknown words, typos, or complex syntax
+
+## How are tokens used?
+
+While training, after a text has been tokenized, each unique token is assigned an ID.
+
+For example, in the sentence:
+
+> I heard a dog bark loudly at a cat
+
+After tokenization the following IDs could be assigned:
+
+- I (1)
+- heard (2)
+- a (3)
+- dog (4)
+- bark (5)
+- loudly (6)
+- at (7)
+- a (the "a" token is already assigned an ID of 3)
+- cat (8)
+
+This allows text to be represented as a sequence of token IDs. The previous sentence would be represented as [1, 2, 3, 4, 5, 6, 7, 3, 8]. The sentence "I heard a cat" would be represented as [1, 2, 3, 8].
+
+As training continues, any new tokens in the training text will be added to the vocabulary and assigned an ID. For example:
+
+- meow (9)
+- run (10)
+
+Using these token ID sequences, the semantic relationships between the tokens can be analyzed. These relationships are represented as multi-valued numeric vectors, known as [_embeddings_](understanding-embeddings.md). Each token will be assigned an embedding based on how commonly they are used together or in similar contexts.
```

Comment:
```suggestion
The semantic relationships between the tokens can be analyzed using these token ID sequences. These relationships are represented as multi-valued numeric vectors, known as [_embeddings_](understanding-embeddings.md). Each token is assigned an embedding based on how commonly it's used together with or in similar contexts to the other tokens in the embedding.
```

Reply:
Not sure if my extrapolation is correct here, but felt the sentence wasn't quite complete.